File: Source/RawParser/Parser/RawSpeedCode/NikonDecompressor.cs
using System;

namespace RawParser
{

    class NikonDecompressor : LJpegDecompressor
    {
        bool uncorrectedRawValues;
        private UInt16[] curve = new UInt16[65536];
        private static byte[][] nikon_tree = {
  { 0,1,5,1,1,1,1,1,1,2,0,0,0,0,0,0,	/* 12-bit lossy */
  5,4,3,6,2,7,1,0,8,9,11,10,12 },
  { 0,1,5,1,1,1,1,1,1,2,0,0,0,0,0,0,	/* 12-bit lossy after split */
  0x39,0x5a,0x38,0x27,0x16,5,4,3,2,1,0,11,12,12 },
  { 0,1,4,2,3,1,2,0,0,0,0,0,0,0,0,0,  /* 12-bit lossless */
  5,4,6,3,7,2,8,1,9,0,10,11,12 },
  { 0,1,4,3,1,1,1,1,1,2,0,0,0,0,0,0,	/* 14-bit lossy */
  5,6,4,7,8,3,9,2,1,0,10,11,12,13,14 },
  { 0,1,5,1,1,1,1,1,1,1,2,0,0,0,0,0,	/* 14-bit lossy after split */
  8,0x5c,0x4b,0x3a,0x29,7,6,5,4,3,2,1,0,13,14 },
  { 0,1,4,2,2,3,1,2,0,0,0,0,0,0,0,0,	/* 14-bit lossless */
  7,6,8,5,9,4,10,3,11,12,2,0,1,13,14 }
        };

        NikonDecompressor(TiffBinary file, RawImage img) : base(file, img)
        {
            for (UInt32 i = 0; i < 0x8000; i++)
            {
                curve[i] = i;
            }
        }

        void initTable(UInt32 huffSelect)
        {
            HuffmanTable* dctbl1 = &huff[0];
            UInt32 acc = 0;
            for (UInt32 i = 0; i < 16; i++)
            {
                dctbl1.bits[i + 1] = nikon_tree[huffSelect][i];
                acc += dctbl1.bits[i + 1];
            }
            dctbl1.bits[0] = 0;

            for (UInt32 i = 0; i < acc; i++)
            {
                dctbl1.huffval[i] = nikon_tree[huffSelect][i + 16];
            }
            createHuffmanTable(dctbl1);
        }

        void DecompressNikon(ByteStream* metadata, UInt32 w, UInt32 h, UInt32 bitsPS, UInt32 offset, UInt32 size)
        {
            UInt32 v0 = metadata.getByte();
            UInt32 v1 = metadata.getByte();
            UInt32 huffSelect = 0;
            UInt32 split = 0;
            int[] pUp1 = new int[2];
            int[] pUp2 = new int[2];
            mUseBigtable = true;

            _RPT2(0, "Nef version v0:%u, v1:%u\n", v0, v1);

            if (v0 == 73 || v1 == 88)
                metadata.skipBytes(2110);

            if (v0 == 70) huffSelect = 2;
            if (bitsPS == 14) huffSelect += 3;

            pUp1[0] = metadata.getShort();
            pUp1[1] = metadata.getShort();
            pUp2[0] = metadata.getShort();
            pUp2[1] = metadata.getShort();

            int _max = 1 << bitsPS & 0x7fff;
            UInt32 step = 0;
            UInt32 csize = metadata.getShort();
            if (csize > 1)
                step = _max / (csize - 1);
            if (v0 == 68 && v1 == 32 && step > 0)
            {
                for (UInt32 i = 0; i < csize; i++)
                    curve[i * step] = metadata.getShort();
                for (int i = 0; i < _max; i++)
                    curve[i] = (curve[i - i % step] * (step - i % step) +
                                curve[i - i % step + step] * (i % step)) / step;
                metadata.setAbsoluteOffset(562);
                split = metadata.getShort();
            }
            else if (v0 != 70 && csize <= 0x4001)
            {
                for (UInt32 i = 0; i < csize; i++)
                {
                    curve[i] = metadata.getShort();
                }
                _max = csize;
            }
            initTable(huffSelect);

            mRaw.whitePoint = curve[_max - 1];
            mRaw.blackLevel = curve[0];
            if (!uncorrectedRawValues)
            {
                mRaw.setTable(curve, _max, true);
            }

            UInt32 x, y;
            BitPumpMSB bits = new BitPumpMSB(mFile, offset, size);
            byte* draw = mRaw.getData();
            UInt16* dest;
            UInt32 pitch = mRaw.pitch;

            int pLeft1 = 0;
            int pLeft2 = 0;
            UInt32 cw = w / 2;
            UInt32 random = bits.peekBits(24);
            //allow gcc to devirtualize the calls below
            RawImageDataU16[] rawdata = mRaw.get();
            for (y = 0; y < h; y++)
            {
                if (split && y == split)
                {
                    initTable(huffSelect + 1);
                }
                dest = (UInt16*)&draw[y * pitch];  // Adjust destination
                pUp1[y & 1] += HuffDecodeNikon(bits);
                pUp2[y & 1] += HuffDecodeNikon(bits);
                pLeft1 = pUp1[y & 1];
                pLeft2 = pUp2[y & 1];
                rawdata.setWithLookUp(clampbits(pLeft1, 15), (byte8*)dest++, &random);
                rawdata.setWithLookUp(clampbits(pLeft2, 15), (byte8*)dest++, &random);
                for (x = 1; x < cw; x++)
                {
                    bits.checkPos();
                    pLeft1 += HuffDecodeNikon(bits);
                    pLeft2 += HuffDecodeNikon(bits);
                    rawdata.setWithLookUp(clampbits(pLeft1, 15), (byte8*)dest++, &random);
                    rawdata.setWithLookUp(clampbits(pLeft2, 15), (byte8*)dest++, &random);
                }
            }

            if (uncorrectedRawValues)
            {
                mRaw.setTable(curve, _max, false);
            }
            else
            {
                mRaw.setTable(null);
            }
        }

        /*
        *--------------------------------------------------------------
        *
        * HuffDecode --
        *
        * Taken from Figure F.16: extract next coded symbol from
        * input stream.  This should becode a macro.
        *
        * Results:
        * Next coded symbol
        *
        * Side effects:
        * Bitstream is parsed.
        *
        *--------------------------------------------------------------
        */
        int HuffDecodeNikon(BitPumpMSB bits)
        {
            int rv;
            int l, temp;
            int code, val;

            HuffmanTable* dctbl1 = &huff[0];

            bits.fill();
            code = bits.peekBitsNoFill(14);
            val = dctbl1.bigTable[code];
            if ((val & 0xff) != 0xff)
            {
                bits.skipBitsNoFill(val & 0xff);
                return val >> 8;
            }

            rv = 0;
            code = bits.peekByteNoFill();
            val = dctbl1.numbits[code];
            l = val & 15;
            if (l)
            {
                bits.skipBitsNoFill(l);
                rv = val >> 4;
            }
            else
            {
                bits.skipBits(8);
                l = 8;
                while (code > dctbl1.maxcode[l])
                {
                    temp = bits.getBitNoFill();
                    code = (code << 1) | temp;
                    l++;
                }

                if (l > 16)
                {
                    throw new Exception("Corrupt JPEG data: bad Huffman code:%u\n", l);
                }
                else
                {
                    rv = dctbl1.huffval[dctbl1.valptr[l] + ((int)(code - dctbl1.((code[l]))];
                }
            }

            if (rv == 16)
                return -32768;

            /*
            * Section F.2.2.1: decode the difference and
            * Figure F.12: extend sign bit
            */
            UInt32 len = rv & 15;
            UInt32 shl = rv >> 4;
            int diff = ((bits.getBits(len - shl) << 1) + 1) << shl >> 1;
            if ((diff & (1 << (len - 1))) == 0)
                diff -= (1 << len) - !shl;
            return diff;
        }
    }
}

File: Source/RawSpeed/decompressors/NikonDecompressor.cpp
/*
    RawSpeed - RAW file decoder.

    Copyright (C) 2009-2014 Klaus Post

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/

#include "decompressors/NikonDecompressor.h"
#include "common/Common.h"                // for uint32, ushort16, clampBits
#include "common/Point.h"                 // for iPoint2D
#include "common/RawImage.h"              // for RawImage, RawImageData, RawI...
#include "decoders/RawDecoderException.h" // for ThrowRDE
#include "decompressors/HuffmanTable.h"   // for HuffmanTable
#include "io/BitPumpMSB.h"                // for BitPumpMSB, BitStream<>::fil...
#include "io/Buffer.h"                    // for Buffer
#include "io/ByteStream.h"                // for ByteStream
#include <cstdio>                         // for size_t, NULL
#include <vector>                         // for vector, allocator

namespace rawspeed {

const uchar8 NikonDecompressor::nikon_tree[][2][16] = {
    {/* 12-bit lossy */
     {0, 1, 5, 1, 1, 1, 1, 1, 1, 2, 0, 0, 0, 0, 0, 0},
     {5, 4, 3, 6, 2, 7, 1, 0, 8, 9, 11, 10, 12}},
    {/* 12-bit lossy after split */
     {0, 1, 5, 1, 1, 1, 1, 1, 1, 2, 0, 0, 0, 0, 0, 0},
     {0x39, 0x5a, 0x38, 0x27, 0x16, 5, 4, 3, 2, 1, 0, 11, 12, 12}},
    {/* 12-bit lossless */
     {0, 1, 4, 2, 3, 1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0},
     {5, 4, 6, 3, 7, 2, 8, 1, 9, 0, 10, 11, 12}},
    {/* 14-bit lossy */
     {0, 1, 4, 3, 1, 1, 1, 1, 1, 2, 0, 0, 0, 0, 0, 0},
     {5, 6, 4, 7, 8, 3, 9, 2, 1, 0, 10, 11, 12, 13, 14}},
    {/* 14-bit lossy after split */
     {0, 1, 5, 1, 1, 1, 1, 1, 1, 1, 2, 0, 0, 0, 0, 0},
     {8, 0x5c, 0x4b, 0x3a, 0x29, 7, 6, 5, 4, 3, 2, 1, 0, 13, 14}},
    {/* 14-bit lossless */
     {0, 1, 4, 2, 2, 3, 1, 2, 0, 0, 0, 0, 0, 0, 0, 0},
     {7, 6, 8, 5, 9, 4, 10, 3, 11, 12, 2, 0, 1, 13, 14}},

};

std::vector<ushort16> NikonDecompressor::createCurve(ByteStream* metadata,
                                                     uint32 bitsPS, uint32 v0,
                                                     uint32 v1, uint32* split) {
  // 'curve' will hold a peace wise linearly interpolated function.
  // there are 'csize' segements, each is 'step' values long.
  // the very last value is not part of the used table but necessary
  // to linearly interpolate the last segment, therefor the '+1/-1'
  // size adjustments of 'curve'.
  std::vector<ushort16> curve((1 << bitsPS & 0x7fff) + 1);
  assert(curve.size() > 1);

  for (size_t i = 0; i < curve.size(); i++)
    curve[i] = i;

  uint32 step = 0;
  uint32 csize = metadata->getU16();
  if (csize > 1)
    step = curve.size() / (csize - 1);

  if (v0 == 68 && v1 == 32 && step > 0) {
    for (size_t i = 0; i < csize; i++)
      curve[i * step] = metadata->getU16();
    for (size_t i = 0; i < curve.size() - 1; i++) {
      curve[i] = (curve[i - i % step] * (step - i % step) +
                  curve[i - i % step + step] * (i % step)) /
                 step;
    }

    metadata->setPosition(562);
    *split = metadata->getU16();
  } else if (v0 != 70) {
    if (csize == 0 || csize > 0x4001)
      ThrowRDE("Don't know how to compute curve! csize = %u", csize);

    curve.resize(csize + 1UL);
    assert(curve.size() > 1);

    for (uint32 i = 0; i < csize; i++) {
      curve[i] = metadata->getU16();
    }
  }

  // and drop the last value
  curve.resize(curve.size() - 1);
  assert(!curve.empty());

  return curve;
}

HuffmanTable NikonDecompressor::createHuffmanTable(uint32 huffSelect) {
  HuffmanTable ht;
  uint32 count = ht.setNCodesPerLength(Buffer(nikon_tree[huffSelect][0], 16));
  ht.setCodeValues(Buffer(nikon_tree[huffSelect][1], count));
  ht.setup(true, false);
  return ht;
}

void NikonDecompressor::decompress(RawImage* mRaw, ByteStream&& data,
                                   ByteStream metadata, const iPoint2D& size,
                                   uint32 bitsPS, bool uncorrectedRawValues) {
  assert(bitsPS > 0);

  uint32 v0 = metadata.getByte();
  uint32 v1 = metadata.getByte();
  uint32 huffSelect = 0;
  uint32 split = 0;
  int pUp1[2];
  int pUp2[2];

  writeLog(DEBUG_PRIO_EXTRA, "Nef version v0:%u, v1:%u", v0, v1);

  if (v0 == 73 || v1 == 88)
    metadata.skipBytes(2110);

  if (v0 == 70)
    huffSelect = 2;
  if (bitsPS == 14)
    huffSelect += 3;

  pUp1[0] = metadata.getU16();
  pUp1[1] = metadata.getU16();
  pUp2[0] = metadata.getU16();
  pUp2[1] = metadata.getU16();

  HuffmanTable ht = createHuffmanTable(huffSelect);

  auto curve = createCurve(&metadata, bitsPS, v0, v1, &split);
  RawImageCurveGuard curveHandler(mRaw, curve, uncorrectedRawValues);

  BitPumpMSB bits(data);
  uchar8* draw = mRaw->get()->getData();
  uint32 pitch = mRaw->get()->pitch;

  int pLeft1 = 0;
  int pLeft2 = 0;
  uint32 random = bits.peekBits(24);
  //allow gcc to devirtualize the calls below
  auto* rawdata = reinterpret_cast<RawImageDataU16*>(mRaw->get());

  assert(size.x % 2 == 0);
  assert(size.x >= 2);
  for (uint32 y = 0; y < static_cast<unsigned>(size.y); y++) {
    if (split && y == split) {
      ht = createHuffmanTable(huffSelect + 1);
    }
    auto* dest =
        reinterpret_cast<ushort16*>(&draw[y * pitch]); // Adjust destination
    pUp1[y&1] += ht.decodeNext(bits);
    pUp2[y&1] += ht.decodeNext(bits);
    pLeft1 = pUp1[y&1];
    pLeft2 = pUp2[y&1];

    rawdata->setWithLookUp(clampBits(pLeft1, 15),
                           reinterpret_cast<uchar8*>(dest + 0), &random);
    rawdata->setWithLookUp(clampBits(pLeft2, 15),
                           reinterpret_cast<uchar8*>(dest + 1), &random);

    dest += 2;

    for (uint32 x = 2; x < static_cast<uint32>(size.x); x += 2) {
      pLeft1 += ht.decodeNext(bits);
      pLeft2 += ht.decodeNext(bits);

      rawdata->setWithLookUp(clampBits(pLeft1, 15),
                             reinterpret_cast<uchar8*>(dest + 0), &random);
      rawdata->setWithLookUp(clampBits(pLeft2, 15),
                             reinterpret_cast<uchar8*>(dest + 1), &random);

      dest += 2;
    }
  }
}

} // namespace rawspeed
